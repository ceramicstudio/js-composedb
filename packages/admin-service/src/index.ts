export { router } from './router.js'
export type { Router } from './router.js'
export { GraphQLQueryCodec, Service } from './service.js'
export type { GraphQLResult, ServiceConfig, ServiceParams } from './service.js'
