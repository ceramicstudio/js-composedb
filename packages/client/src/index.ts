/**
 * High-level ComposeDB client, based on the {@linkcode runtime ComposeDB runtime}.
 *
 * ## Installation
 *
 * ```sh
 * npm install @composedb/client
 * ```
 *
 * @module client
 */

export { ComposeClient, type ComposeClientParams } from './client.js'
