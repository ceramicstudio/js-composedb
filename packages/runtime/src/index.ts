/**
 * ComposeDB runtime module, converting a runtime composite to an executable GraphQL schema.
 *
 * ## Installation
 *
 * ```sh
 * npm install @composedb/runtime
 * ```
 *
 * @module runtime
 */

export { ComposeRuntime, type ComposeRuntimeParams } from './runtime.js'
export { type Context, type ContextParams, createContext } from './context.js'
export type { DocumentCache } from './loader.js'
export { createGraphQLSchema } from './schema.js'
export { getSchema, printGraphQLSchema } from './utils.js'
export type { GetSchemaParams } from './utils.js'
