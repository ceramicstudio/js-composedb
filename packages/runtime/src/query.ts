import type { AnyValueFilter, ObjectFilter, QueryFilters } from '@ceramicnetwork/common'

const GREATER_THAN_CONDITIONS = ['greaterThan', 'greaterThanOrEqualTo']
const LESS_THAN_CONDITIONS = ['lessThan', 'lessThanOrEqualTo']
export function areValidRangeValueConditions(a: string, b: string): boolean {
  return (
    (GREATER_THAN_CONDITIONS.includes(a) && LESS_THAN_CONDITIONS.includes(b)) ||
    (GREATER_THAN_CONDITIONS.includes(b) && LESS_THAN_CONDITIONS.includes(a))
  )
}

export function assertValidAnyValueFilter(
  filter: Record<string, unknown>,
): asserts filter is AnyValueFilter {
  if (filter == null) {
    throw new Error('No value filter provided')
  }

  const keys = Object.keys(filter)
  if (keys.length === 0) {
    throw new Error('Invalid empty value filter')
  }
  if (keys.length > 2) {
    throw new Error('Too many conditions provided in value filter')
  }
  if (keys.length === 2) {
    if (areValidRangeValueConditions(keys[0], keys[1])) {
      return
    }
    throw new Error('Invalid range conditions in value filter')
  }

  const key = keys[0]
  const value = filter[key]
  switch (key) {
    case 'in':
    case 'notIn':
      if (!Array.isArray(value)) {
        throw new Error(`Invalid value for condition "${key}": must be an array`)
      }
      return
    case 'isNull':
    case 'equalTo':
    case 'notEqualTo':
      return
    default:
      if (GREATER_THAN_CONDITIONS.includes(key) || LESS_THAN_CONDITIONS.includes(key)) {
        return
      }
      throw new Error(`Unexpected condition "${key}" in value filter`)
  }
}

export function assertValidObjectFilter(
  filter: Record<string, Record<string, unknown>>,
): asserts filter is ObjectFilter {
  if (filter == null) {
    throw new Error('No object filter provided')
  }

  const values = Object.values(filter)
  if (values.length === 0) {
    throw new Error('Empty object filter')
  }
  values.forEach(assertValidAnyValueFilter)
}

export function assertValidQueryFilters(
  filters: Record<string, unknown>,
): asserts filters is QueryFilters {
  if (filters == null) {
    throw new Error('No query filters provided')
  }

  const keys = Object.keys(filters)
  if (keys.length !== 1) {
    throw new Error('Invalid query filters: a single key must be provided')
  }

  const key = keys[0]
  const value = filters[key]
  if (value == null) {
    throw new Error('Invalid query filters: missing filter value')
  }

  switch (key) {
    case 'where':
      assertValidObjectFilter(value as Record<string, Record<string, unknown>>)
      return
    case 'not':
      assertValidQueryFilters(value as Record<string, Record<string, unknown>>)
      return
    case 'and':
    case 'or':
      if (!Array.isArray(value)) {
        throw new Error(`Invalid value for "${key}" filter: must be an array`)
      }
      value.forEach(assertValidQueryFilters)
      return
    default:
      throw new Error(
        `Invalid query filter type: expecting "where", "and", "or" or "not" key but got "${key}"`,
      )
  }
}

export function createRelationQueryFilters(
  relationKey: string,
  targetID: string,
  customFilters?: QueryFilters,
): QueryFilters {
  const filters: QueryFilters = {
    where: { [relationKey]: { equalTo: targetID } },
  }
  if (customFilters == null) {
    return filters
  }

  assertValidQueryFilters(customFilters)
  return { and: [filters, customFilters] }
}
