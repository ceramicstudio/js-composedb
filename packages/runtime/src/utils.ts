import type { RuntimeCompositeDefinition } from '@composedb/types'
import { type GraphQLSchema, printSchema } from 'graphql'

import { createGraphQLSchema } from './schema.js'

export type GetSchemaParams = {
  /**
   * Runtime composite definition, created using the {@linkcode devtools.Composite Composite}
   * development tools.
   */
  definition?: RuntimeCompositeDefinition
  /**
   * Set the schema to read-only, disabling mutations support.
   */
  readonly?: boolean
  /**
   * GraphQL Schema to use, ignores the `definition` and `readonly` parameters if provided.
   */
  schema?: GraphQLSchema
}

/**
 * Use or create a GraphQL schema based on the provided parameters.
 */
export function getSchema(params: GetSchemaParams): GraphQLSchema {
  const { definition, readonly, schema } = params
  if (schema != null) {
    return schema
  }
  if (definition == null) {
    throw new Error(`Invalid parameters to get schema: missing definition or schema`)
  }
  return createGraphQLSchema({ definition, readonly })
}

/**
 * Create a GraphQL schema from a runtime composite definition and return its string
 * representation.
 */
export function printGraphQLSchema(
  definition: RuntimeCompositeDefinition,
  readonly = false
): string {
  return printSchema(createGraphQLSchema({ definition, readonly }))
}
