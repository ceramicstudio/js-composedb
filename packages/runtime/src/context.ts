import type { BaseQuery, CeramicApi } from '@ceramicnetwork/common'
import type { ModelInstanceDocument } from '@ceramicnetwork/stream-model-instance'
import type { CommitID, StreamID } from '@ceramicnetwork/streamid'
import type { Connection } from 'graphql-relay'

import { type ConnectionQuery, queryConnection, querySingle, toBaseQuery } from './query.js'
import { type DocumentCache, DocumentLoader, type UpdateDocOptions } from './loader.js'

export type ContextParams = {
  /**
   * Optional cache for documents.
   */
  cache?: DocumentCache | boolean
  /**
   * Ceramic client instance.
   */
  ceramic: CeramicApi
  /**
   * Fallback viewer ID to use when the Ceramic instance is not authenticated.
   */
  fallbackViewerID?: string | null
  /**
   * @internal
   */
  loader?: DocumentLoader
}

export type Context = {
  /**
   * Returns whether the Ceramic client instance used internally is authenticated or not. When not
   * authenticated, mutations will fail.
   */
  isAuthenticated: () => boolean
  /**
   * Ceramic client instance used internally.
   */
  ceramic: CeramicApi
  /**
   * Document loader instance used internally.
   */
  loader: DocumentLoader
  /**
   * ID of the current viewer (authenticated DID), if set.
   */
  getViewerID: () => string | null
  /**
   * Load a document by ID, using the cache if possible.
   */
  loadDoc: <Content extends Record<string, any>>(
    id: string | CommitID | StreamID,
    fresh?: boolean
  ) => Promise<ModelInstanceDocument<Content>>
  /**
   * Create a new document with the given model and content.
   */
  createDoc: <Content extends Record<string, any>>(
    model: string,
    content: Content
  ) => Promise<ModelInstanceDocument<Content>>
  /**
   * Create a new single document with the given model and content.
   */
  createSingle: <Content extends Record<string, any>>(
    model: string,
    content: Content
  ) => Promise<ModelInstanceDocument<Content>>
  /**
   * Update an existing document.
   */
  updateDoc: <Content extends Record<string, any>>(
    id: string | StreamID,
    content: Content,
    options?: UpdateDocOptions
  ) => Promise<ModelInstanceDocument<Content>>
  /**
   * Query the index for a connection of documents.
   */
  queryConnection: (query: ConnectionQuery) => Promise<Connection<ModelInstanceDocument | null>>
  /**
   * Query the index for the total number of documents matching the query parameters.
   */
  queryCount: (query: BaseQuery) => Promise<number>
  /**
   * Query the index for a single document.
   */
  querySingle: (query: BaseQuery) => Promise<ModelInstanceDocument | null>
}

export function createContext(params: ContextParams): Context {
  const { cache, ceramic, fallbackViewerID } = params
  const loader = params.loader ?? new DocumentLoader({ ceramic, cache })

  function getViewerID(): string | null {
    const did = ceramic.did
    return did ? (did.hasParent ? did.parent : did.id) : fallbackViewerID ?? null
  }

  return {
    ceramic,
    loader,
    getViewerID,
    isAuthenticated: (): boolean => ceramic.did?.authenticated ?? false,
    loadDoc: async <Content extends Record<string, any> = Record<string, any>>(
      id: string | CommitID | StreamID,
      fresh = false
    ): Promise<ModelInstanceDocument<Content>> => {
      if (fresh) {
        loader.clear(id)
      }
      return await loader.load<Content>(id)
    },
    createDoc: async <Content extends Record<string, any> = Record<string, any>>(
      model: string,
      content: Content
    ): Promise<ModelInstanceDocument<Content>> => {
      return await loader.create(model, content)
    },
    createSingle: async <Content extends Record<string, any> = Record<string, any>>(
      model: string,
      content: Content
    ): Promise<ModelInstanceDocument<Content>> => {
      const controller = getViewerID()
      if (controller == null) {
        throw new Error('Document can only be created with an authenticated account')
      }
      const doc = await loader.single<Content>(controller, model)
      await doc.replace(content)
      return doc
    },
    updateDoc: async <Content extends Record<string, any> = Record<string, any>>(
      id: string | StreamID,
      content: Content,
      options?: UpdateDocOptions
    ): Promise<ModelInstanceDocument<Content>> => {
      return await loader.update(id, content, options)
    },
    queryConnection: async (
      query: ConnectionQuery
    ): Promise<Connection<ModelInstanceDocument | null>> => {
      return await queryConnection(ceramic, query)
    },
    queryCount: async (query: BaseQuery): Promise<number> => {
      return await ceramic.index.count(toBaseQuery(query))
    },
    querySingle: async (query: BaseQuery): Promise<ModelInstanceDocument | null> => {
      return await querySingle(ceramic, query)
    },
  }
}
