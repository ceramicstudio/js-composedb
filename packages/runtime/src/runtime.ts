import type { CeramicApi } from '@ceramicnetwork/common'
import { CeramicClient } from '@ceramicnetwork/http-client'
import {
  type DocumentNode,
  type ExecutionResult,
  type GraphQLError,
  type Source,
  GraphQLSchema,
  execute,
  parse,
  validate,
} from 'graphql'

import { type Context, createContext } from './context.js'
import type { DocumentCache } from './loader.js'
import { type GetSchemaParams, getSchema } from './utils.js'

export type ComposeRuntimeParams = GetSchemaParams & {
  /**
   * Optional cache for documents.
   */
  cache?: DocumentCache | boolean
  /**
   * Ceramic client instance or HTTP URL.
   */
  ceramic: CeramicApi | string
  /**
   * Optional context to use.
   */
  context?: Context
}

/**
 * The ComposeRuntime class provides APIs to execute queries on a GraphQL schema generated from a
 * `RuntimeCompositeDefinition`. It allows applications to interact with documents using known
 * models on a Ceramic node.
 *
 * It is exported by the {@linkcode runtime} module.
 *
 * ```sh
 * import { ComposeRuntime } from '@composedb/runtime'
 * ```
 */
export class ComposeRuntime {
  #context: Context
  #schema: GraphQLSchema

  constructor(params: ComposeRuntimeParams) {
    const { ceramic, context, definition, readonly, schema, ...contextParams } = params
    const ceramicClient = typeof ceramic === 'string' ? new CeramicClient(ceramic) : ceramic
    this.#context = context ?? createContext({ ...contextParams, ceramic: ceramicClient })
    this.#schema = getSchema({ definition, readonly, schema })
  }

  /**
   * Context object used internally.
   */
  get context(): Context {
    return this.#context
  }

  /**
   * Execute a GraphQL query from a DocumentNode and optional variables.
   */
  async execute<Data = Record<string, unknown>>(
    document: DocumentNode,
    variableValues?: Record<string, unknown>
  ): Promise<ExecutionResult<Data>> {
    const errors = validate(this.#schema, document)
    return errors.length > 0
      ? { errors }
      : ((await execute({
          document,
          variableValues,
          contextValue: this.#context,
          schema: this.#schema,
        })) as unknown as ExecutionResult<Data>)
  }

  /**
   * Execute a GraphQL query from its source and optional variables.
   */
  async executeQuery<Data = Record<string, unknown>>(
    source: string | Source,
    variableValues?: Record<string, unknown>
  ): Promise<ExecutionResult<Data>> {
    let document: DocumentNode
    try {
      document = parse(source)
    } catch (syntaxError) {
      return { errors: [syntaxError as GraphQLError] }
    }
    return await this.execute<Data>(document, variableValues)
  }
}
