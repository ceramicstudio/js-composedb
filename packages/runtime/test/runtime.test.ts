/**
 * @jest-environment composedb
 */

import type { CeramicApi } from '@ceramicnetwork/common'
import { CommitID, StreamID } from '@ceramicnetwork/streamid'
import { Composite } from '@composedb/devtools'
import {
  createCommentSchemaWithPost,
  loadPostSchemaWithComments,
  noteSchema,
  postSchema,
  profilesSchema,
  extraScalarsSchema,
} from '@composedb/test-schemas'
import { AccountId, ChainId } from 'caip'
import { CID } from 'multiformats/cid'

import { ComposeRuntime, printGraphQLSchema } from '../src'

declare global {
  const ceramic: CeramicApi
}

describe('runtime', () => {
  test('create profile', async () => {
    const composite = await Composite.create({ ceramic, schema: profilesSchema })
    const runtime = new ComposeRuntime({ ceramic, definition: composite.toRuntime() })
    const res = await runtime.executeQuery<{ createGenericProfile: { document: { id: string } } }>(
      `
      mutation CreateProfile($input: CreateGenericProfileInput!) {
        createGenericProfile(input: $input) {
          document {
            id
          }
        }
      }
      `,
      { input: { content: { name: 'Alice' } } }
    )
    expect(res.data?.createGenericProfile.document.id).toBeDefined()
  }, 30000)

  test('create and query post with comments', async () => {
    const postComposite = await Composite.create({ ceramic, schema: postSchema })
    const postModelID = postComposite.modelIDs[0]

    const commentComposite = await Composite.create({
      ceramic,
      schema: createCommentSchemaWithPost(postModelID),
    })
    const commentModelID = commentComposite.modelIDs.find((id) => id !== postModelID) as string

    const composite = await Composite.create({
      ceramic,
      schema: loadPostSchemaWithComments(postModelID, commentModelID),
    })
    const definition = composite.toRuntime()

    expect(printGraphQLSchema(definition)).toMatchSnapshot()

    const runtime = new ComposeRuntime({ ceramic, definition })

    const createPostMutation = `
      mutation CreatePost($input: CreatePostInput!) {
        createPost(input: $input) {
          document {
            id
          }
        }
      }
    `
    await runtime.executeQuery(createPostMutation, {
      input: { content: { title: 'A first post', text: 'First post content' } },
    })
    const postRes = await runtime.executeQuery<{ createPost: { document: { id: string } } }>(
      createPostMutation,
      { input: { content: { title: 'A second post', text: 'Second post content' } } }
    )
    const postID = postRes.data?.createPost.document.id

    const createCommentMutation = `
      mutation CreateComment($input: CreateCommentInput!) {
        createComment(input: $input) {
          document {
            id
          }
        }
      }
    `
    await runtime.executeQuery(createCommentMutation, {
      input: { content: { postID, text: 'A first comment' } },
    })
    await runtime.executeQuery(createCommentMutation, {
      input: { content: { postID, text: 'A second comment' } },
    })
    await runtime.executeQuery(createCommentMutation, {
      input: { content: { postID, text: 'A third comment' } },
    })

    const res = await runtime.executeQuery(
      `
      query {
        viewer {
          postList(first: 5) {
            edges {
              node {
                title
                text
                commentsCount
                comments(first: 5) {
                  edges {
                    node {
                      text
                    }
                  }
                }
              }
            }
          }
        }
      }
      `
    )
    expect(res).toMatchSnapshot()
  }, 60000)

  test('can create a document using extra scalars', async () => {
    const composite = await Composite.create({ ceramic, schema: extraScalarsSchema })
    const runtime = new ComposeRuntime({ ceramic, definition: composite.toRuntime() })
    const res = await runtime.executeQuery<{ createExtraScalars: { document: { id: string } } }>(
      `
      mutation CreateScalars($input: CreateExtraScalarsInput!) {
        createExtraScalars(input: $input) {
          document {
            accountID
            chainID
            cid
            commitID
            countryCode
            date
            dateTime
            did {
              id
            }
            latitude
            localDate
            locale
            localTime
            longitude
            streamID
            time
            timeZone
            uri
            utcOffset
          }
        }
      }
      `,
      {
        input: {
          content: {
            accountID: new AccountId('eip155:1:0xabcdef'),
            chainID: new ChainId('eip155:1'),
            cid: CID.parse('bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a'),
            commitID: CommitID.fromString(
              'k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju'
            ),
            countryCode: 'US',
            date: '2023-01-31',
            dateTime: '2023-01-31T16:04:16.475Z',
            did: 'did:test:123',
            latitude: 53.471,
            localDate: '2023-01-31',
            locale: 'en-gb',
            localTime: '14:25:06.123',
            longitude: 53.471,
            streamID: StreamID.fromString(
              'kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s'
            ),
            time: '14:10:20+01:00',
            // TimeZone seem to fail in CI
            // timeZone: 'America/Costa_Rica',
            uri: 'https://ceramic.network',
            utcOffset: '+01:15',
          },
        },
      }
    )
    expect(res.data?.createExtraScalars.document).toMatchSnapshot()
  }, 20000)

  test('create a document with enum', async () => {
    const composite = await Composite.create({ ceramic, schema: noteSchema })
    const runtime = new ComposeRuntime({ ceramic, definition: composite.toRuntime() })
    const res = await runtime.executeQuery<{ createNote: { document: { id: string } } }>(
      `
      mutation CreateNote($input: CreateNoteInput!) {
        createNote(input: $input) {
          document {
            status
            title
            text
          }
        }
      }
      `,
      {
        input: {
          content: {
            status: 'DEFAULT',
            title: 'A test note',
            text: 'Test node contents',
          },
        },
      }
    )
    expect(res.data?.createNote.document).toMatchSnapshot()
  }, 20000)
})
