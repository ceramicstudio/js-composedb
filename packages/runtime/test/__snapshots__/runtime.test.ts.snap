// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runtime can create a document using extra scalars 1`] = `
{
  "accountID": "eip155:1:0xabcdef",
  "chainID": "eip155:1",
  "cid": "bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a",
  "commitID": "k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju",
  "countryCode": "US",
  "date": "2023-01-31",
  "dateTime": "2023-01-31T16:04:16.475Z",
  "did": {
    "id": "did:test:123",
  },
  "latitude": 53.471,
  "localDate": "2023-01-31",
  "localTime": "14:25:06.123",
  "locale": "en-gb",
  "longitude": 53.471,
  "streamID": "kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s",
  "time": "14:10:20+01:00",
  "timeZone": null,
  "uri": "https://ceramic.network",
  "utcOffset": "+01:15",
}
`;

exports[`runtime create a document with enum 1`] = `
{
  "status": "DEFAULT",
  "text": "Test node contents",
  "title": "A test note",
}
`;

exports[`runtime create and query post with comments 1`] = `
"type Query {
  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node

  """Account currently authenticated on the Ceramic instance, if set"""
  viewer: CeramicAccount
  commentIndex(
    """Returns the items in the list that come after the specified cursor."""
    after: String

    """Returns the first n items from the list."""
    first: Int

    """Returns the items in the list that come before the specified cursor."""
    before: String

    """Returns the last n items from the list."""
    last: Int
  ): CommentConnection
  postIndex(
    """Returns the items in the list that come after the specified cursor."""
    after: String

    """Returns the first n items from the list."""
    first: Int

    """Returns the items in the list that come before the specified cursor."""
    before: String

    """Returns the last n items from the list."""
    last: Int
  ): PostConnection
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

type CeramicAccount implements Node {
  """Globally unique identifier of the account (DID string)"""
  id: ID!

  """
  Whether the Ceramic instance is currently authenticated with this account or not
  """
  isViewer: Boolean!
  commentList(
    """Returns the items in the list that come after the specified cursor."""
    after: String

    """Returns the first n items from the list."""
    first: Int

    """Returns the items in the list that come before the specified cursor."""
    before: String

    """Returns the last n items from the list."""
    last: Int
  ): CommentConnection
  postList(
    """Returns the items in the list that come after the specified cursor."""
    after: String

    """Returns the first n items from the list."""
    first: Int

    """Returns the items in the list that come before the specified cursor."""
    before: String

    """Returns the last n items from the list."""
    last: Int
  ): PostConnection
}

"""A connection to a list of items."""
type CommentConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [CommentEdge]
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""An edge in a connection."""
type CommentEdge {
  """The item at the end of the edge"""
  node: Comment

  """A cursor for use in pagination"""
  cursor: String!
}

type Comment implements Node {
  id: ID!
  text: String!
  postID: CeramicStreamID!
  post: Post

  """Account controlling the document"""
  author: CeramicAccount!

  """Current version of the document"""
  version: CeramicCommitID!
}

"""A Ceramic Stream ID"""
scalar CeramicStreamID @specifiedBy(url: "https://github.com/ceramicnetwork/CIP/blob/main/CIPs/CIP-59/CIP-59.md#streamid")

type Post implements Node {
  id: ID!
  date: DateTime
  text: String!
  title: String!

  """Account controlling the document"""
  author: CeramicAccount!

  """Current version of the document"""
  version: CeramicCommitID!
  commentsCount(
    """Counts only documents created by the provided account"""
    account: ID
  ): Int!
  comments(
    """Returns the items in the list that come after the specified cursor."""
    after: String

    """Returns the first n items from the list."""
    first: Int

    """Returns the items in the list that come before the specified cursor."""
    before: String

    """Returns the last n items from the list."""
    last: Int

    """Returns only documents created by the provided account"""
    account: ID
  ): CommentConnection!
}

"""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
"""
scalar DateTime

"""A Ceramic Commit ID"""
scalar CeramicCommitID @specifiedBy(url: "https://github.com/ceramicnetwork/CIP/blob/main/CIPs/CIP-59/CIP-59.md#commitid")

"""A connection to a list of items."""
type PostConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [PostEdge]
}

"""An edge in a connection."""
type PostEdge {
  """The item at the end of the edge"""
  node: Post

  """A cursor for use in pagination"""
  cursor: String!
}

type Mutation {
  createComment(input: CreateCommentInput!): CreateCommentPayload
  updateComment(input: UpdateCommentInput!): UpdateCommentPayload
  createPost(input: CreatePostInput!): CreatePostPayload
  updatePost(input: UpdatePostInput!): UpdatePostPayload
}

type CreateCommentPayload {
  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node

  """Account currently authenticated on the Ceramic instance, if set"""
  viewer: CeramicAccount
  document: Comment!
  clientMutationId: String
}

input CreateCommentInput {
  content: CommentInput!
  clientMutationId: String
}

input CommentInput {
  text: String!
  postID: CeramicStreamID!
}

type UpdateCommentPayload {
  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node

  """Account currently authenticated on the Ceramic instance, if set"""
  viewer: CeramicAccount
  document: Comment!
  clientMutationId: String
}

input UpdateCommentInput {
  id: ID!
  content: PartialCommentInput!
  options: UpdateOptionsInput
  clientMutationId: String
}

input PartialCommentInput {
  text: String
  postID: CeramicStreamID
}

input UpdateOptionsInput {
  """
  Fully replace the document contents instead of performing a shallow merge
  """
  replace: Boolean = false

  """Only perform mutation if the document matches the provided version"""
  version: CeramicCommitID
}

type CreatePostPayload {
  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node

  """Account currently authenticated on the Ceramic instance, if set"""
  viewer: CeramicAccount
  document: Post!
  clientMutationId: String
}

input CreatePostInput {
  content: PostInput!
  clientMutationId: String
}

input PostInput {
  date: DateTime
  text: String!
  title: String!
}

type UpdatePostPayload {
  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node

  """Account currently authenticated on the Ceramic instance, if set"""
  viewer: CeramicAccount
  document: Post!
  clientMutationId: String
}

input UpdatePostInput {
  id: ID!
  content: PartialPostInput!
  options: UpdateOptionsInput
  clientMutationId: String
}

input PartialPostInput {
  date: DateTime
  text: String
  title: String
}"
`;

exports[`runtime create and query post with comments 2`] = `
{
  "data": {
    "viewer": {
      "postList": {
        "edges": [
          {
            "node": {
              "comments": {
                "edges": [],
              },
              "commentsCount": 0,
              "text": "First post content",
              "title": "A first post",
            },
          },
          {
            "node": {
              "comments": {
                "edges": [
                  {
                    "node": {
                      "text": "A first comment",
                    },
                  },
                  {
                    "node": {
                      "text": "A second comment",
                    },
                  },
                  {
                    "node": {
                      "text": "A third comment",
                    },
                  },
                ],
              },
              "commentsCount": 3,
              "text": "Second post content",
              "title": "A second post",
            },
          },
        ],
      },
    },
  },
}
`;
